docs/csharp/programming-guide/classes-and-structs/local-functions.md

```diff
@@ -80,22 +80,39 @@ Like with the method iterator, you can refactor the preceding example and put th
+### Typing
 
-Third, the compiler can perform static analysis that enables local functions to definitely assign captured variables in the enclosing scope. Consider this example:
+Lambda expressions rely on the typing of the `Func`/`Action` variable that they're assigned to determine the argument and return types. In local functions, since the syntax is much like writing a normal method, argument types and return type are already part of the function declaration.
+
+### Definite Assignment
+
+Local functions have different rules for definite assignment than lambda expressions. A local function declaration can be referenced from any code location where it is in scope; in our first example `LocalFunctionFactorial`, we could declare our local function either above or below the `return` statement and not trigger any compiler errors.
+
+Lambda expressions follow the rules for definite assignment more closely. In order for a lambda expression to be used, the `Func`/`Action` variable that it will be assigned to must be declared and the lambda expression assigned to it. Notice that `LambdaFactorial` must declare and initialize the lambda expression `nthFactorial` before defining it. Not doing so results in a compile time error for referencing `nthFactorial` before assigning it.
+
+These differences mean that recursive algorithms are easier to create using local functions. You can declare and define a local function that calls itself. Lambda expressions must be declared, and assigned a default value before they can be re-assigned to a body that references the same lambda expression.
+
+### Conversion to a Delegate
+
+Both local functions and lambda expressions require that any captured variables are definitely assigned when either is converted to a delegate, but *when* they're converted differs. Lambda expressions are converted to delegates when they're declared, while local functions are only converted to delegates when ***used*** as a delegate.
+
+If you declare a local function and only reference it by calling it like a method, it will not be converted to a delegate. That rule enables you to declare a local function at any convenient location in its enclosing scope. It's common to declare local functions at the end of the parent method, after any return statements.
```

Comment:
These two paragraphs have a number of subtle and misleading statements.  As written, I don't have a great suggestion, but here are the issues with it. 

- Its mixing a discussion of the definite assignment rules for captured variables and conversion to a delegate. 
- It's an implementation detail whether or not a local function is implemented as a delegate. (When it happens, the delegate definition is a compiler generated name with private scope, and can't be referenced as such from source code). 
- The sentence on declaration location is covered better above where you make the distinction between a function definition and a lambda declaration and assignment.

I think this H3 can be removed. If you chose that route, I'd add a sentence in the next section on variable capture. Namely, that local functions are implemented as delegate types when they capture variables in the enclosing scope.

Reply:
I've been sitting on this part for a while. This was the hardest part of the section to parse when I was reading the current live docs since I believe there were multiple topics being discussed in the same paragraph. I *think* having the H3 is beneficial so long as I'm accurately covering what you've specified, so I'm toying with the following:

---

### Definite assignment

Lambda expressions are objects that are declared and assigned at runtime. In order for a lambda expression to be used, it needs to be [definitely assigned](../../../../_csharplang/spec/variables.md#definite-assignment): the `Func`/`Action` variable that it will be assigned to must be declared and the lambda expression assigned to it. Notice that `LambdaFactorial` must declare and initialize the lambda expression `nthFactorial` before defining it. Not doing so results in a compile time error for referencing `nthFactorial` before assigning it.

Local functions are defined at compile time. As they're not assigned to variables, they can be referenced from any code location **where it is in scope**; in our first example `LocalFunctionFactorial`, we could declare our local function either above or below the `return` statement and not trigger any compiler errors.

These differences mean that recursive algorithms are easier to create using local functions. You can declare and define a local function that calls itself. Lambda expressions must be declared, and assigned a default value before they can be re-assigned to a body that references the same lambda expression.

### Variable capture

The rules of [definite assignment](../../../../_csharplang/spec/variables.md#definite-assignment) also affect any variables that are captured by the local function or lambda expression. The compiler can perform static analysis that enables local functions to definitely assign captured variables in the enclosing scope. Consider this example:

```csharp
// code
```

The compiler can determine that `LocalFunction` definitely assigns `y` when called. Because `LocalFunction` is called before the `return` statement, `y` is definitely assigned at the `return` statement.

Note that when a local function captures variables in the enclosing scope, the local function is implemented as a delegate type in the eyes of the compiler.

---

The trailing sentence on declaration location is removed and there's no mixing of definite assignment and delegate conversion. However, I'm not entirely sure how to handle the delegate conversion part since it's an implementation detail. This comment is already getting pretty long but I'm not sure if the following should stay as its own section or if it should be moved into the `Variable capture` section:

---

### Implementation as a delegate

Lambda expressions are converted to delegates when they're declared. Local functions are more flexible in that they have the *option* to be implemented as a delegate. Local functions are only converted to delegates when they're used as delegates. If you declare a local function and only reference it by calling it like a method, it will not be converted to a delegate.